src/mlir/design.md

```diff
@@ -0,0 +1,158 @@
+        ...
+      }
+    }
+  }
+}
+```
+
+This structure, however, does not convert naturally to LLVM dialect. So we have
+a later pass that mangles the names and moves all functions to the root module
+just before lowering to LLVM dialect. The example above would be:
+
+```
+module @__ {
+  func @____FooBar__Foo__foo() {
+    ...
+  }
+}
+```
+
+And all calls to that function would also be changed in every other function that
+calls this one.
+
+## Load/Store Semantics
+
+Stack variables are allocated by LLVM IR with an `alloca` instruction. MLIR has
+a similar concept and we use that in the first implementation. The main problem
+with this approach is that there is only one dialect that has alloca/load/store:
+`memref`.
+
+MemRefs represent the memory used by `tensors`, which are multi-dimensional
+objects (matrices) of the same type. Verona is obviously not restricted to
+tensors as types, so we'll need to extend this to structure types, with complex
+layout semantics (ex. union types). For that, we'll need a set of dialect
+instructions that can be later lowered directly into LLVM for allocating stack
+space on complex layout semantics.
+
+For now, every function argument and local declaration (`let`, `var`) are
+allocated in the stack and its value stored in the memory space. This simplifies
+code generation (all variable symbols point to an address), but it doesn't help
+with value symbols and are really slow.
+
+The LLVM pipeline, even at O1, recognises most of the argument, temporary and
+local allocas as bogus and optimises them away, so this isn't a critical issue
+from a performance point of view, but we will probably need to update the symbol
+table soon to account for more types and states of the variables we hold (ex.
+`let` vs `var`).
+
+## Runtime Calls
+
+The Verona runtime is where the magic happens. Allocating regions, choosing the
+memory management strategy, scheduling behaviours, dispatching foreign language
+code, etc. The AST, however, has only the semantics of the language (with hints
+to the runtime behaviour), so we need to translate it at the MLIR lowering level.
+
+For example, heap allocation, which LLVM uses `malloc` calls, Verona uses
+`snmalloc`, which has a finer control over what we can do with the regions. In
+the end, however, it will be the same: calls to the runtime library in IR.
+
+The generator knows what runtime calls are appropriate for each AST construct,
+insert those calls in the right places and makes sure to declare the methods
+so that linking with the runtime library works.
+
+## Arithmetic
+
+Arithmetic is generated by the generator via pattern matching on the types and
+"method" names. Adding two numbers (`a + b`) is the same as calling the method
+`+` on `a` over `b` (`a.+(b)`), but these methods don't exist (how would we
+implement addition if not with a `+`?).
+
+So the generator has to find those that weren't declared and replace with
+arithmetic if the types are compatible (int, float).
```

Comment:
Smalltalk handled this with ['primitive' methods](http://wiki.squeak.org/squeak/5716): The standard library `#new` method on `Object`, for example, invokes the `#basicNew' method on the class.  It would be good to discuss why we're not going down this path.

Reply:
I honestly don't know enough about Smalltalk (or language design) to start this discussion, nor I think this is the right document to do so. This document is about the MLIR lowering, not the language design.

The AST doesn't have a `primitive` attribute and, IIUC, we don't want one to exist, so we can create functions with the same names on different types without resorting to detecting what's primitive and what's not.

Reply:
Primitive methods are not really part of the language in Smalltalk, they're a property of the implementation.  The basic idea is that the implementation reserves a few selectors as things that are not subject to dynamic dispatch and for which implementations can't be provided.  Applying this to Verona, we'd pick a prefix like `$primitive` and have a `$primitiveAdd` selector that would let you do `x $primitiveAdd y` where `x` and `y` are numeric types.  You'd then implement `+` in the standard library as a method that just calls `$primitiveAdd` on the receiver and the argument.  The `+` method would then appear in vtables (if necessary for any numeric types cast to interfaces), and would in every other way behave just like a normal method (because it is).  The `$primitiveAdd` method would be lowered here.  Implementing a `$primitive{X}` method would be a parse error.  There's then a clear contract between the compiler and the standard library that a set of `$primitive` methods must exist on certain types.

Reply:
I think I understood it like you describe, in a way that's what we're doing, but there is no `primitiveAdd` in the AST and I believe this is intentional (@sylvanc?).

Numeric types in the AST are `class U32 {}` types, which in the AST are still classes that just happen to have that name. I am not sure if we want users to override numeric behaviour by creating their own `U32` classes instead of `using "numeric"`, but the current AST would allow that (we have tests that do that). In those cases, `+` could do whatever and it wouldn't be a `primitiveAdd` because it would have declared a '+' function in that context and we would have found it before trying arithmetic.

I believe Pony has a late binding of arithmetic at the LLVM IR lowering stage, once it knows there isn't anything else that overrides the behaviour (I'm not sure about the Pony semantics here). In Verona, I'm doing it a lot earlier, but also checking to see we don't have anything else defined (which should be fine for now).

Specifically for the MLIR lowering, being `+` with a late match or being `primitiveAdd` upfront is irrelevant. So, if we want to discuss the merits of each implementation, we need to look at the parser, not the MLIR lowering.